utils/ujson-merge.man
.\" Manpage for ujson-verify
.\" Contact dan@ultramarin.se to correct errors or types.
.TH ujson-merge 1 "" "" "User Commands"


.SH NAME
ujson-merge \- Merge multiple JSON documents into one.


.SH SYNOPSIS
.B ujson-merge
[OPTIONS] MASTER_FILE [SLAVE_FILE...]


.SH DESCRIPTION
ujson-merge joins one or more JSON documents into one master JSON document and prints the result to standard output.
If only a master document is present on the command line, it will be merged from standard input.
.PP
When merging JSON instance B into JSON instance A, the following rules apply:
.nf
- If A is a number, it will be replaced by the instance B.
- If A is null, it will be replaced by the instance B.
- If A is a boolean, it will be replaced by the instance B.
- If A is a string, it will be replaced by the instance B.
- If A is an array, it will be replaced by the instance B.
- If A is an object and B is not an object, A will be replaced by instance B.
- If both A and B are objects:
    - If a member name in B exitst in A, then B will be merged into A using theese rules.
    - If a member name in B doesn't exitst A, that member will be added to A,


.SH OPTIONS
.TP
.B -c, --compact
Print the resulting JSON document without whitespaces.
.TP
.B -r, --relaxed
Parse JSON documents in relaxed mode.

.TP
.B -d, --dst-pointer=POINTER
Merge into this specific place in the master document.

.TP
.B -s, --src-pointer=POINTER
Merge from this specific place in the slave document(s).

.TP
.B -m, --match-type
All values that are merged must be of the same type. Note that the content in two arrays may be of different types.

.TP
.B -h, --help
Print help and exit.


.SH POINTER
A JSON Pointer as described in RFC 6901.
A pointer specifies a specific value in a JSON document. Is is a string containing zero or more reference tokens separated by '/'. If a '/' is part of a token it is encodec as '~1'. And if a '~' is parf of a token it is encoded as '~0'.


.SH EXAMPLES
Given the following JSON documents:
.PP
Document A.json:
.EX
{
    "name": "master",
    "value": 0,
    "list": [1, 2, 3],
    "obj": {
        "desc": null,
        "place": "home"
    }
}
.EE
.PP
Document B.json:
.EX
{
    "name": "replacement",
    "obj": {
        "desc": "Hello World!",
        "misc": [true, false]
    },
    "meaning": 42
}
.EE

.pp
Result of runnning command "ujson-merge A.json B.json":
.EX
{
    "name": "replacement",
    "value": 0,
    "list": [1, 2, 3],
    "obj": {
        "desc": "Hello World!",
        "place": "home",
        "misc": [true, false]
    },
    "meaning": 42
}
.EE

.pp
Result of runnning command "ujson-merge --match-type A.json B.json":
.EX
JSON type mismatch. Trying to assign a value of type string to an instance of type null
.EE

.pp
Result of runnning command "ujson-merge --dst-pointer=/list --src-pointer=/obj/misc A.json B.json":
.EX
{
    "name": "master",
    "value": 0,
    "list": [true, false],
    "obj": {
        "desc": null,
        "place": "home"
    }
}
.EE


.SH SEE ALSO
ujson-verify(1) ujson-cmp(1) ujson-print(1) ujson-get(1)


.SH AUTHOR
Dan Arrhenius (dan@ultramarin.se)


utils/ujson-merge.cpp
/*
 * Copyright (C) 2021 Dan Arrhenius <dan@ultramarin.se>
 *
 * This file is part of ujson.
 *
 * ujson is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published
 * by the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
#include <ujson.hpp>
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <algorithm>
#include <unistd.h>
#include <getopt.h>


using namespace std;

static constexpr const char* prog_name = "ujson-merge";

struct appargs_t {
    bool compact;
    bool relaxed;
    bool type_check;
    string master_filename;
    vector<string> slave_filenames;
    string master_pointer;
    string slave_pointer;

    appargs_t () {
        compact = false;
        relaxed = false;
        type_check = false;
    }
};


/*
 * Merge rules:
 *
 * Merge a slave instance into a master instance.
 *
 * A number will be replaced by the slave value.
 * A null will be replaced by the slave value.
 * A boolean will be replaced by the slave value.
 * A string will be replaced by the slave value.
 * An array will be replaced by the slave value.
 * An object:
 *   - If a member name in the slave object exitst in the master object,
 *     the value of that member will be merged according to theese rules.
 *   - If a member name in the slave object doesn't exitst in the master object,
 *     the slave member will be added to the master object.
 */


//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
static void print_usage_and_exit (std::ostream& out, int exit_code)
{
    out << endl;
    out << "Merge one or more JSON document into a target document.." << endl;
    out << endl;
    out << "Usage: " << prog_name << " [OPTIONS] [MASTER_FILE] [SLAVE_FILE...]" << endl;
    out << endl;
    out << "MASTER_FILE: This JSON document serves as the base for the merge." <<endl;
    out << "SLAVE_FILE:  Values from this file will be merged with the master file." <<endl;
    out << "The resulting JSON document will be printed to standard output." << endl;
    out << endl;
    out << "Options:" <<endl;
    out << "  -c, --compact              Print the resulting JSON document without whitespaces." << endl;
    out << "  -r, --relaxed              Parse JSON documents in relaxed mode." << endl;
    out << "  -d, --dst-pointer=POINTER  Merge into this specific place in the master document." << endl;
    out << "  -s, --src-pointer=POINTER  Merge from this specific place in the slave document(s)." << endl;
    out << "  -m, --match-type           The value(s) in the master and slave document must be of the same type." << endl;
    out << "  -h, --help                 Print this help message and exit." << endl;
    out << endl;
    out << "" << endl;
    exit (exit_code);
}


//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
static void parse_args (int argc, char* argv[], appargs_t& args)
{
    struct option long_options[] = {
        { "compact",      no_argument,       0, 'c'},
        { "relaxed",      no_argument,       0, 'r'},
        { "dst-pointer",  required_argument, 0, 'd'},
        { "src-pointer",  required_argument, 0, 's'},
        { "match-type",   no_argument,       0, 'm'},
        { "help",         no_argument,       0, 'h'},
        { 0, 0, 0, 0}
    };
    const char* arg_format = "crd:s:mh";

    while (1) {
        int c = getopt_long (argc, argv, arg_format, long_options, NULL);
        if (c == -1)
            break;
        switch (c) {
        case 'c':
            args.compact = true;
            break;
        case 'r':
            args.relaxed = true;
            break;
        case 'd':
            args.master_pointer = optarg;
            break;
        case 's':
            args.slave_pointer = optarg;
            break;
        case 'm':
            args.type_check = true;
            break;
        case 'h':
            print_usage_and_exit (std::cout, 0);
            break;
        default:
            exit (1);
            break;
        }
    }
    if (optind < argc) {
        args.master_filename = argv[optind++];
    }else{
        cerr << "Missing argument" << endl;
        exit (1);
    }
    while (optind < argc)
        args.slave_filenames.emplace_back (argv[optind++]);
}


//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
static int merge_instance (ujson::jvalue& mvalue, ujson::jvalue& svalue, bool type_check)
{
    // Type check
    if (type_check && mvalue.type()!=svalue.type()) {
        cerr << "JSON type mismatch. Trying to assign a value of type "
             << ujson::jtype_to_str(svalue.type())
             << " to an instance of type " << ujson::jtype_to_str(mvalue.type()) << endl;
        return 1;
    }
    // Sanity check
    if (mvalue.type()!=ujson::j_invalid && svalue.type()==ujson::j_invalid) {
        cerr << "JSON type mismatch. Trying to assign an invalid JSON value "
             << " to an instance of type " << ujson::jtype_to_str(mvalue.type()) << endl;
        return 1;
    }

    if (mvalue.type()!=ujson::j_object || svalue.type()!=ujson::j_object) {
        mvalue = svalue;
    }else{
        for (auto& sv : svalue.obj()) {
            auto& mv = mvalue.get(sv.first);
            if (mv.valid()) {
                if (merge_instance(mv, sv.second, type_check))
                    return 1;
            }else{
                mvalue.add (sv.first, sv.second);
            }
        }
    }

    return 0;
}


//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
static int parse_and_merge (const appargs_t& opt,
                            ujson::jvalue& master,
                            istream& ifs,
                            const std::string& slave_filename="")
{
    ujson::Json j;
    string slave_document ((istreambuf_iterator<char>(ifs)), istreambuf_iterator<char>());
    auto slave_root = j.parse_string (slave_document, !opt.relaxed);
    if (!slave_root.valid()) {
        if (slave_filename.empty())
            cerr << "Parse error on input: " << j.error() << endl;
        else
            cerr << "Parse error, " << slave_filename << ": " << j.error() << endl;
        return 1;
    }

    auto& slave = ujson::find_jvalue (slave_root, opt.slave_pointer);
    if (!slave.valid()) {
        if (slave_filename.empty())
            cerr << "Invalid pointer in input document: " << opt.slave_pointer << endl;
        else
            cerr << "Invalid pointer in document "
                 << slave_filename << ": " << opt.slave_pointer << endl;
        return 1;
    }
    return merge_instance (master, slave, opt.type_check);
}


//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
int main (int argc, char* argv[])
{
    appargs_t opt;
    parse_args (argc, argv, opt);

    // Parse the master file
    //
    ujson::Json j;
    auto master_root = j.parse_file (opt.master_filename, !opt.relaxed);
    if (!master_root.valid()) {
        cerr << "Parse error, " << opt.master_filename << ": " << j.error() << endl;
        exit (1);
    }

    // Get the pointer in the master instance
    //
    auto& master = ujson::find_jvalue (master_root, opt.master_pointer);
    if (!master.valid()) {
        cerr << "Invalid pointer in document "
             << opt.master_filename << ": " << opt.master_pointer << endl;
        exit (1);
    }

    // Merge slave file(s)
    //
    int result = 0;
    if (opt.slave_filenames.empty()) {
        // Merge from standard input
        result = parse_and_merge (opt, master, cin);
    }else{
        // Merge from file(s)
        for (auto& filename : opt.slave_filenames) {
            ifstream ifs (filename);
            result = parse_and_merge (opt, master, ifs, filename);
            ifs.close ();
            if (result)
                break;
        }
    }

    if (!result)
        cout << master_root.describe(!opt.compact, !opt.relaxed) << endl;

    return result;
}
